src/lib/tls/msg_server_hello.cpp

```diff
@@ -662,6 +776,44 @@ Protocol_Version Server_Hello_13::selected_version() const
 Hello_Retry_Request::Hello_Retry_Request(std::unique_ptr<Server_Hello_Internal> data)
    : Server_Hello_13(std::move(data), Server_Hello_13::as_hello_retry_request) {}
 
+Hello_Retry_Request::Hello_Retry_Request(const Client_Hello_13& ch, Named_Group selected_group, const Policy& policy, Callbacks& cb)
+   : Server_Hello_13(std::make_unique<Server_Hello_Internal>(
+                        Protocol_Version::TLS_V12 /* legacy_version */,
+                        ch.session_id(),
+                        HELLO_RETRY_REQUEST_MARKER,
+                        choose_ciphersuite(ch, policy),
+                        uint8_t(0) /* compression method */,
+                        true /* is Hello Retry Request */
+                     ), as_new_hello_retry_request)
+   {
+   // RFC 8446 4.1.4
+   //     As with the ServerHello, a HelloRetryRequest MUST NOT contain any
+   //     extensions that were not first offered by the client in its
+   //     ClientHello, with the exception of optionally the "cookie" [...]
+   //     extension.
+   BOTAN_STATE_CHECK(ch.extensions().has<Supported_Groups>());
+   BOTAN_STATE_CHECK(ch.extensions().has<Key_Share>());
```

Comment:
Yes, albeit somewhat implicit, when processing the Client Hello in `tls_server_impl_13.cpp`. For now we must just assume that both extensions are availabe, because we don't support any PSK-mode in the server. Once PSK-based negotiation (particularly without ephemeral key exchange) is being implemented we will need to re-iterate this.